src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,121 @@
+    private static final int ELEMENT_SHIFT_STEP = 1;
+    private static final int HALF_CAPACITY = 2;
+    private static final int ZERO_POSITION = 0;
+    private T[] elements;
+    private int size;
+
+    @SuppressWarnings("unchecked")
+    public ArrayList() {
+        elements = (T[]) new Object[DEFAULT_CAPACITY];
+        size = 0;
+    }
+
     @Override
     public void add(T value) {
-
+        ensureCapacity();
+        elements[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkIndexForAdd(index);
+        if (size == elements.length) {
+            growSize();
+        }
+        for (int i = size; i > index; i--) {
+            elements[i] = elements[i - 1];
+        }
+        elements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return (T) elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T value = elements[index];
+        manualArrayCopy(elements, index + ELEMENT_SHIFT_STEP, elements,
+                index, size - index - ELEMENT_SHIFT_STEP);
+        elements[--size] = null;
+        return value;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if ((element == null && elements[i] == null)
+                    || (element != null && element.equals(elements[i]))) {
+                return remove(i);
+            }
+        }
+        throw new NoSuchElementException("Element not found");
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void growSize() {
+        int newCapacity = elements.length + elements.length / HALF_CAPACITY;
+        T[] newElements = (T[]) new Object[newCapacity];
+        manualArrayCopy(elements, ZERO_POSITION,
+                newElements, ZERO_POSITION, size);
+        elements = newElements;
+    }
+
+    private void ensureCapacity() {
+        if (size >= elements.length) {
+            growSize();
+        }
+    }
+
+    private void manualArrayCopy(T[] source, int sourcePos, T[] dest, int destPos, int length) {
+        for (int i = 0; i < length; i++) {
+            dest[destPos + i] = source[sourcePos + i];
+        }
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Index: " + index
+                    + " is out of bounds");
```

Comment:
include size in the message, makes it easier to debug, you can see that you're off by one for example